### PR TITLE
[GEOS-10324] adding arbitrary number of columns to ogcLink tables

### PR DESCRIPTION
**Pull request by ydeliorman**
[GEOS-10324](https://osgeo-org.atlassian.net/browse/GEOS-10324)

OgcLink tables in OSEO now support adding an arbitrary number of columns.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/community/oseo/oseo-core/src/test/resources/postgis.sql [hunk 1]
```diff
@@ -206,7 +206,10 @@ create table collection_ogclink (
   "method" varchar,
   "code" varchar,
   "type" varchar,
-  "href" varchar
+  "href" varchar,
```
Comment (aaime) [inline src/community/oseo/oseo-core/src/test/resources/postgis.sql:209]: Nope... the ``postgis.sql`` is linked from the documentation as the reference structure for the database. We cannot have testing bits in it.
That said, you can run ``alter table`` to add columns in ``collection_test_data.sql`` and ``product_test_data.sql``, those files are test only instead.

File: src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/AbstractOpenSearchController.java [hunk 2]
```diff
@@ -217,7 +223,35 @@ protected OgcLinks buildOgcLinksFromFeature(Feature feature, boolean notFoundOnE
         List<OgcLink> links = Collections.emptyList();
         Collection<Property> linkProperties =
                 feature.getProperties(OpenSearchAccess.OGC_LINKS_PROPERTY_NAME);
-        if (linkProperties != null) {
+
+        Map<String, String> unknownFields = new HashMap<>();
+        // taken from OgcLink properties. these properties are known(static) from OgcLink.
+        Set<String> knownFields =
```
Comment (aaime) [inline src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/AbstractOpenSearchController.java:229]: This is a constant, please don't declare it in methods, but at the top of the file as a proper constant (static final, all uppercase name).

File: src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/AbstractOpenSearchController.java [hunk 3]
```diff
@@ -217,7 +223,35 @@ protected OgcLinks buildOgcLinksFromFeature(Feature feature, boolean notFoundOnE
...
+                    (SimpleFeatureImpl) ((ArrayList) linkProperties).get(0);
+
+            // iterate over feature to get unknown properties from OgcLink
+            List<AttributeType> types = simpleFeature.getType().getTypes();
+            for (int i = 0; i < types.size(); i++) {
+                String unknownColumnName = types.get(i).getName().getLocalPart();
+                if (!knownFields.contains(unknownColumnName)
+                        && simpleFeature.getAttributes().get(i) != null) {
+                    unknownFields.put(
+                            unknownColumnName, simpleFeature.getAttributes().get(i).toString());
```
Comment (aaime) [inline src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/AbstractOpenSearchController.java:251]: Hum... string? No, the data type from the database should be preserved:
- When asking for the JSON representation of OGCLink, I expect numbers to be non-quoted, as well as booleans, and dates formatted as ISO dates.
- This code is not null safe, a property can legitimately be null.

Please change your test database to include columns with types int, float, boolean and date. Make sure the types are respected as the data flows across the layers of code, and in the REST API, as well as getting reported properly in templates (JSON templates is type aware too).

File: src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/OgcLink.java [hunk 4]
```diff
@@ -20,16 +25,25 @@
 
     String href;
 
+    Map<String, String> unknownFields = new HashMap<>();
```
Comment (aaime) [inline src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/OgcLink.java:28]: I believe it should be ``Map<String, Object>``, see comments above about respecting types.

**Review by aaime — CHANGES_REQUESTED** [review]
Looks like a pretty decent start, but does not respect the data types in the database. Please fix.

File: src/community/oseo/oseo-core/src/test/resources/postgis.sql (hunk 1 above)
Comment (ydeliorman) [reply src/community/oseo/oseo-core/src/test/resources/postgis.sql]: restored to the original state. added add column queries to SQL files provided

File: src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/AbstractOpenSearchController.java (hunk 2 above)
Comment (ydeliorman) [reply src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/AbstractOpenSearchController.java]: done

File: src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/AbstractOpenSearchController.java (hunk 3 above)
Comment (ydeliorman) [reply src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/AbstractOpenSearchController.java]: Changed to the object type. Dates are written in ISO date format.

File: src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/OgcLink.java (hunk 4 above)
Comment (ydeliorman) [reply src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/OgcLink.java]: done

**Comment by ydeliorman** [comment]
review is done.

File: src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/UnknownFieldsSerializer.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package org.geoserver.opensearch.rest;
```
Comment (aaime) [inline src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/UnknownFieldsSerializer.java:1]: Lacks standard copyright header

**Review by aaime — CHANGES_REQUESTED** [review]

File: src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/UnknownFieldsSerializer.java (hunk 5 above)
Comment (ydeliorman) [reply src/community/oseo/oseo-rest/src/main/java/org/geoserver/opensearch/rest/UnknownFieldsSerializer.java:5]: Added the same copyright of [EncodingHints](https://github.com/geoserver/geoserver/blob/26273424f084bbee64e0e7e672307b8fa8deb03c/src/community/features-templating/features-templating-core/src/main/java/org/geoserver/featurestemplating/builders/EncodingHints.java#L1-L4).

**Review by aaime — APPROVED** [review]

**Pull request merged**